### ISSUE TITLE
Unnecessary netty-all lib on classpath + inconsistency between netty versions

### ISSUE THREAD

**Issue by jovanstevanovic** [renaissance-benchmarks/renaissance#337]
This is the list of all netty libraries on the classpath for apache spark benchmark:

- benchmarks/apache-spark/netty-all-4.1.68.Final.jar
- benchmarks/apache-spark/netty-buffer-4.1.50.Final.jar
- benchmarks/apache-spark/netty-codec-4.1.50.Final.jar
- benchmarks/apache-spark/netty-common-4.1.50.Final.jar
- benchmarks/apache-spark/netty-handler-4.1.50.Final.jar
- benchmarks/apache-spark/netty-resolver-4.1.50.Final.jar
- benchmarks/apache-spark/netty-transport-4.1.50.Final.jar
- benchmarks/apache-spark/netty-transport-native-epoll-4.1.50.Final.jar
- benchmarks/apache-spark/netty-transport-native-unix-common-4.1.50.Final.jar

And I think it's not necessary to keep both netty-all and netty-* libs on the classpath. 

The second problem is that apache-spark is using two different netty versions 4.10.50 (for netty-* libs) and 4.10.68 (for netty-all). This could lead to a problems similar to the one I found:
 
> java.lang.NoSuchMethodError: 'void io.netty.buffer.PooledByteBufAllocator.<init>(boolean, int, int, int, int, int, int, boolean)'
> 	at org.apache.spark.network.util.NettyUtils.createPooledByteBufAllocator(NettyUtils.java:171)
> 	at org.apache.spark.network.util.NettyUtils.getSharedPooledByteBufAllocator(NettyUtils.java:142)
> 	at org.apache.spark.network.client.TransportClientFactory.<init>(TransportClientFactory.java:111)
> 	at org.apache.spark.network.TransportContext.createClientFactory(TransportContext.java:142)
> 	at org.apache.spark.rpc.netty.NettyRpcEnv.<init>(NettyRpcEnv.scala:77)
> 	at org.apache.spark.rpc.netty.NettyRpcEnvFactory.create(NettyRpcEnv.scala:493)
> 	at org.apache.spark.rpc.RpcEnv$.create(RpcEnv.scala:57)
> 	at org.apache.spark.SparkEnv$.create(SparkEnv.scala:268)
> 	at org.apache.spark.SparkEnv$.createDriverEnv(SparkEnv.scala:191)
> 	at org.apache.spark.SparkContext.createSparkEnv(SparkContext.scala:277)
> 	at org.apache.spark.SparkContext.<init>(SparkContext.scala:460)
> 	at org.apache.spark.SparkContext$.getOrCreate(SparkContext.scala:2690)
> 	at org.apache.spark.sql.SparkSession$Builder.$anonfun$getOrCreate$2(SparkSession.scala:949)
> 	at scala.Option.getOrElse(Option.scala:201)

In this example, the implementation of `io.netty.buffer.PooledByteBufAllocator` has been changed between 0.50 and 0.68. `NettyUtils` (from `spark-network-common_2.13-3.2.0`) is expecting `ByteBufAllocator` from .68, but received it from 0.50.

**Comment by lbulej** [comment]
Wow, that's a good catch, thanks! The libs on the classpath are the result of normal transitive dependency resolution done by SBT/coursier, so there must be something wrong along the path (we don't really tweak anything except ask for `spark-core`, `spark-sql` and `spark-mllib`).  I'll try to see what we can do about that.

**Comment by jovanstevanovic** [comment]
I'm glad I could help! Let me know once you find out something :smile:

**Comment by farquet** [comment]
The `sbt dependencyTree` command shows those relevant bits from within the `apache-spark` sbt project:

```
[info]     | +-org.apache.zookeeper:zookeeper:3.6.2
[info]     | | +-commons-lang:commons-lang:2.6
[info]     | | +-io.netty:netty-handler:4.1.50.Final
[info]     | | | +-io.netty:netty-buffer:4.1.50.Final
[info]     | | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | |
[info]     | | | +-io.netty:netty-codec:4.1.50.Final
[info]     | | | | +-io.netty:netty-buffer:4.1.50.Final
[info]     | | | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | | |
[info]     | | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | | +-io.netty:netty-transport:4.1.50.Final
[info]     | | | |   +-io.netty:netty-buffer:4.1.50.Final
[info]     | | | |   | +-io.netty:netty-common:4.1.50.Final
[info]     | | | |   |
[info]     | | | |   +-io.netty:netty-common:4.1.50.Final
[info]     | | | |   +-io.netty:netty-resolver:4.1.50.Final
[info]     | | | |     +-io.netty:netty-common:4.1.50.Final
[info]     | | | |
[info]     | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | +-io.netty:netty-resolver:4.1.50.Final
[info]     | | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | |
[info]     | | | +-io.netty:netty-transport:4.1.50.Final
[info]     | | |   +-io.netty:netty-buffer:4.1.50.Final
[info]     | | |   | +-io.netty:netty-common:4.1.50.Final
[info]     | | |   |
[info]     | | |   +-io.netty:netty-common:4.1.50.Final
[info]     | | |   +-io.netty:netty-resolver:4.1.50.Final
[info]     | | |     +-io.netty:netty-common:4.1.50.Final
[info]     | | |
[info]     | | +-io.netty:netty-transport-native-epoll:4.1.50.Final
[info]     | | | +-io.netty:netty-buffer:4.1.50.Final
[info]     | | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | |
[info]     | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | +-io.netty:netty-transport-native-unix-common:4.1.50.Final
[info]     | | | | +-io.netty:netty-buffer:4.1.50.Final
[info]     | | | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | | |
[info]     | | | | +-io.netty:netty-common:4.1.50.Final
[info]     | | | | +-io.netty:netty-transport:4.1.50.Final
[info]     | | | |   +-io.netty:netty-buffer:4.1.50.Final
[info]     | | | |   | +-io.netty:netty-common:4.1.50.Final
[info]     | | | |   |
[info]     | | | |   +-io.netty:netty-common:4.1.50.Final
[info]     | | | |   +-io.netty:netty-resolver:4.1.50.Final
[info]     | | | |     +-io.netty:netty-common:4.1.50.Final
[info]     | | | |
[info]     | | | +-io.netty:netty-transport:4.1.50.Final
[info]     | | |   +-io.netty:netty-buffer:4.1.50.Final
[info]     | | |   | +-io.netty:netty-common:4.1.50.Final
[info]     | | |   |
[info]     | | |   +-io.netty:netty-common:4.1.50.Final
[info]     | | |   +-io.netty:netty-resolver:4.1.50.Final
[info]     | | |     +-io.netty:netty-common:4.1.50.Final
```
and 
```
[info]     | +-org.apache.spark:spark-network-shuffle_2.13:3.2.0
[info]     | | +-io.dropwizard.metrics:metrics-core:4.2.0
[info]     | | | +-org.slf4j:slf4j-api:1.7.30
[info]     | | |
[info]     | | +-org.apache.spark:spark-network-common_2.13:3.2.0 [S]
[info]     | | | +-com.fasterxml.jackson.core:jackson-annotations:2.12.3
[info]     | | | +-com.fasterxml.jackson.core:jackson-databind:2.12.3
[info]     | | | | +-com.fasterxml.jackson.core:jackson-annotations:2.12.3
[info]     | | | | +-com.fasterxml.jackson.core:jackson-core:2.12.3
[info]     | | | |
[info]     | | | +-com.google.code.findbugs:jsr305:3.0.0 (evicted by: 3.0.2)
[info]     | | | +-com.google.code.findbugs:jsr305:3.0.2
[info]     | | | +-com.google.crypto.tink:tink:1.6.0
[info]     | | | | +-com.google.code.gson:gson:2.8.6
[info]     | | | | +-com.google.protobuf:protobuf-java:3.14.0
[info]     | | | |
[info]     | | | +-io.dropwizard.metrics:metrics-core:4.2.0
[info]     | | | | +-org.slf4j:slf4j-api:1.7.30
[info]     | | | |
[info]     | | | +-io.netty:netty-all:4.1.68.Final
[info]     | | | +-org.apache.commons:commons-crypto:1.1.0
[info]     | | | +-org.apache.commons:commons-lang3:3.12.0
[info]     | | | +-org.fusesource.leveldbjni:leveldbjni-all:1.8
[info]     | | | +-org.roaringbitmap:RoaringBitmap:0.9.0
[info]     | | | | +-org.roaringbitmap:shims:0.9.0
```

**Comment by lbulej** [comment]
This should be fixed by the [0.14.0 release](https://github.com/renaissance-benchmarks/renaissance/releases/tag/v0.14.0). Feel free to reopen if the problem persists.

**Comment by jovanstevanovic** [comment]
Great! I'll let you know if anything happens 😄
